content/tutorial-escrow.md

```diff
@@ -19,6 +19,406 @@ When testing in [stand-alone mode](concept-stand-alone-mode.html), you can force
+
+*Websocket*
+
+```json
+{% include 'code_samples/escrow/websocket/tx-response-escrowfinish-time.json' %}
+```
+
+<!-- MULTICODE_BLOCK_END -->
+
+
+## Send a conditionally-held escrow
+
+### 1. Generate condition and fulfillment
+
+XRP Ledger escrows require PREIMGE-SHA-256 [Crypto-Conditions](https://tools.ietf.org/html/draft-thomas-crypto-conditions-03). To calculate a condition and fulfillment in the proper format, you should use a Crypto-Conditions library such as [five-bells-condition](https://github.com/interledgerjs/five-bells-condition). For fulfillments, Ripple recommends using one of the following methods to generate the fulfillment:
+
+- Use a cryptographically secure source of randomness to generate at least 32 random bytes
+- Follow Interledger Protocol's [PSK specification](https://github.com/interledger/rfcs/blob/master/0016-pre-shared-key/0016-pre-shared-key.md) and use an HMAC-SHA-256 of the ILP packet as the fulfillment.
+
+Example JavaScript code for a random fulfillment and condition:
+
+```js
+cc = require('five-bells-condition');
+
+const fulfillment_bytes = crypto.randomBytes(32);
+const myFulfillment = new cc.PreimageSha256();
+myFulfillment.setPreimage(fulfillment_bytes);
+console.log(myFulfillment.serializeBinary().toString('hex'));
+// (Random hexadecimal, 72 chars in length)
+console.log(myFulfillment.getConditionBinary().toString('hex'));
+// (Random hexadecimal, 78 chars in length)
+```
+
+Save the condition and the fulfillment for later. Be sure to keep the fulfillment secret until you want to finish executing the held payment; anyone who knows the fulfillment can finish the escrow, releasing the held funds to their intended destination.
+
+
+### 2. Calculate release or cancel time
+
+A Conditional `Escrow` transaction must contain either a `CancelAfter` or `FinishAfter` field, or both. The `CancelAfter` field lets the XRP revert to the sender if the condition is not fulfilled before the specified time. The `FinishAfter` field specifies a time before which the escrow cannot execute, even if someone sends the correct fulfillment. Whichever field you provide, the time it specifies must be in the future.
+
+Example for setting a `CancelAfter` time of 24 hours in the future:
+
+<!-- MULTICODE_BLOCK_START -->
+
+_JavaScript_
+
+```js
+const rippleOffset = 946684800;
+const CancelAfter = Math.floor(Date.now() / 1000) + (24*60*60) - rippleOffset;
+console.log(CancelAfter);
+// Example: 556927412
+```
+
+_Python 2/3_
+
+```python
+from time import time
+ripple_offset = 946684800
+cancel_after = int(time()) + (24*60*60) - 946684800
+print(cancel_after)
+# Example: 556927412
+```
+
+<!-- MULTICODE_BLOCK_END -->
+
+**Warning:** In the XRP Ledger, you must specify time as seconds since the Ripple Epoch (2000-01-01T00:00:00Z). If you use a UNIX time in the `CancelAfter` or `FinishAfter` field without converting to the equivalent Ripple time first, that sets the unlock time to an extra **30 years** in the future!
```

Comment:
Parallelize 'seconds since the Ripple Epoch' in bold:

> **Warning:** In the XRP Ledger, you must specify time as **seconds since the Ripple Epoch** (2000-01-01T00:00:00Z). If you use a UNIX time in the `CancelAfter` or `FinishAfter` field without converting to the equivalent Ripple time first, that sets the unlock time to an extra **30 years** in the future!